core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4266,18 +4266,29 @@ protected RestartCause() {
     @RequirePOST
     public void doExit( StaplerRequest req, StaplerResponse rsp ) throws IOException {
         checkPermission(ADMINISTER);
-        LOGGER.severe(String.format("Shutting down VM as requested by %s from %s",
-                getAuthentication().getName(), req!=null?req.getRemoteAddr():"???"));
-        if (rsp!=null) {
-            rsp.setStatus(HttpServletResponse.SC_OK);
-            rsp.setContentType("text/plain");
-            try (PrintWriter w = rsp.getWriter()) {
-                w.println("Shutting down");
-            }
-        }
+        new Thread("exit thread") {
+            @Override
+            public void run() {
+                try {
+                    ACL.impersonate(ACL.SYSTEM);
+                    LOGGER.severe(String.format("Shutting down VM as requested by %s from %s",
+                            getAuthentication().getName(), req!=null?req.getRemoteAddr():"???"));
+                    if (rsp!=null) {
```

Comment:
I think all of the stuff that modifies the response needs to happen outside of the thread after `start` is called. Otherwise, the method will have already returned, so I'm not sure what would even happen if you modify the response.